### PR TITLE
[WIP -- do not merge] rework 'you may cast|play' one shot effects 

### LINKED ISSUE #10576: Opposition Agent Bugged
Played Opposition Agent alongside Maralen of the Mornsong. Exiled one card from each of my opponent's libraries. Opposition Agent was stolen by a Gilded Drake. I then Defiled it. I no longer had access to the cards I exiled with it which goes against the text on Opposition Agent. My opponent played their own Opposition Agent and then ***played the cards I had exiled previously with my Opposition Agent***. Either I'm *really* not understanding a rules interaction or Opposition Agent is bugged.

### PR DESCRIPTION
**Pull request by Susucre**
Here is some ground work to make sure effects properly distinguish play/cast inside one shot effects. (#10565 )
`CardUtil.makeCardPlayableOrCastable` is now attempting to be a centralized path for those kind of effect.
Not sure how to call those really, those are the 'you may play|cast XXX until end of turn', so neither the static effects (like future sight), nor the 'you may play|cast XXX right now' one shot effects.
For instance Hostage Taker, Gonti Lord of Luxury, and Ragavan Nimble Pilferer are among the reworked cards.

There is a few subeffects of the method not yet coded. I plan to tackle that and add unit tests soon.

In addition to play/cast the three frequent payment adjustment have been set on a card-by-card basis with an enum,
with null having the meaning that none should apply.
- "you may spend mana as though it were mana of any color"
- "you may spend mana as though it were mana of any type"
- "you may (cast | play) ... without paying its mana cost"

Most card with such an effect should have been covered. Hopefully I did that without introducing new bugs. 

A list of cards that would deserve to use the centralized methods, but need a little more work to do so:
- [ ] Arvinox, the Mind Flail: cast only permanents
- [ ] Chandra, Hope's Beacon: cast one instant/sorcery
- [ ] The Omenkeel: play only lands
- [ ] Flameskull: only one
- [ ] Furious Rise: effect discarded on next trigger
- [ ] Havengul Lich: trigger on casting with the special asThought
- [ ] Intet the Dreamer: effect stops on inter leaving the battlefield.
- [ ] Nivix, Aerie of the Firemind: effect checks instant/sorcery to apply. But after that any spell could be cast?
- [ ] Temporal Aperture: stops when card is no longer the revealed one.
- [ ] Unlucky Witness: one of them only.
- [ ] Meeting of the Five: cast only cards that are exactly 3 colors.
- [ ] Xander's Pact: pay life instead of mana cost. 
- [ ] Nashi, Moon Sage's Scion: pay life instead of mana cost. 
- [ ] Fires of Mount Doom: trigger on using the AsThoughtEffect. 
- [ ] Lobelia, Defender of Bag End: cast one among exiled by Lobelia. Not sure this one should be in scope.

Refactor without any particular difficulties/keeping up with newly added cards:
- [ ] Capricious Sliver
- [ ] Chandra, Legacy of Fire
- [ ] Elkin Bottle
- [ ] Grinning Totem
- [ ] Silas Renn, Seeker Adept
- [ ] Primordial Mist
- [ ] Muse Vessel
- [ ] Galvanic Relay
- [ ] Emry, Lurker of the Loch
- [ ] Lidless Gaze
- [ ] Yasmin Khan
- [ ] The Flux

Keeping a list of cards that are probably bugged in the current code, and I may succesfully fix them here:

- [ ] Nahiri, Forged in Fury, in relation to mdfc Equipment / non Equipment.
- [ ] Narset, Enlightned Master, in relation to mdfc creature / non-creature
- [x] Opposition Agent #10576

- [ ] investigate what's up with https://github.com/magefree/mage/pull/11114#discussion_r1335213002 

- [ ] check Devouring Intellect #11205

Not yet in the codebase:
- Chiss-Goria, Forge Tyrant: cast one. Add affinity to artifact. 
- Commander Liara Portier
- Durnan of the Yawning Portal: add undaunted.
- Tlincalli Hunter
- Prowl, Stoic Strategist 
- Kayla's Music Box
- The Ruinous Powers: trigger on cast 

fix #10576 (I'll write a few tests for that one)

### CODE REVIEW

**Comment by xenohedron** [comment]
Failing tests for reference. Some could just be due to choice being handled differently, probably some of these are in fact broken though.

I marked the PR as draft (when opening a PR that is still WIP, you can open it as draft)

```
[ERROR] Failures: 
[ERROR]   PlayFromNonHandZoneTest.castAdventureWithFallenShinobiTest:467->CardTestPlayerAPIImpl.execute:283 Must have playable ability, but not found: Cast Angelic Purge
[ERROR]   PlayFromNonHandZoneTest.castFromExileButWithAdditionalCost2Test:407->CardTestPlayerAPIImpl.execute:283 Can't find ability to activate command: Cast Angelic Purge
[ERROR]   PlayFromNonHandZoneTest.castFromExileButWithAdditionalCostTest:356->CardTestPlayerAPIImpl.execute:283 Can't find ability to activate command: Cast Tormenting Voice
[ERROR]   PlayFromNonHandZoneTest.castSunderingWithGolos2Test:307->CardTestPlayerAPIImpl.execute:283 Can't find ability to activate command: Cast Silvercoat Lion
[ERROR]   PlayFromNonHandZoneTest.castSunderingWithGolosTest:259->CardTestPlayerAPIImpl.execute:283 Can't find ability to activate command: Cast Silvercoat Lion
[ERROR]   PlayFromNonHandZoneTest.testNarsetEnlightenedMaster:110->CardTestPlayerAPIImpl.execute:283 Can't find ability to activate command: Cast Abzan Banner
[ERROR]   PlayFromNonHandZoneTest.testNarsetEnlightenedMasterAdditionalCost:156->CardTestPlayerAPIImpl.execute:283 Must have playable ability, but not found: Cast Cathartic Reunion
[ERROR]   PlayFromNonHandZoneTest.testNarsetEnlightenedMasterAdditionalCost_AI:190->CardTestPlayerAPIImpl.assertGraveyardCount:1455 (Graveyard PlayerA) Card counts are not equal (Swamp) expected:<1> but was:<0>
[ERROR]   PlayTopCardFromLibraryTest.test_MindsDesire_CreaturePlay:142->CardTestPlayerAPIImpl.execute:283 Can't find ability to activate command: Cast Balduvian Bears
[ERROR]   PlayTopCardFromLibraryTest.test_MindsDesire_SplitLeftPlay:214->CardTestPlayerAPIImpl.execute:283 Can't find ability to activate command: Cast Revival$target=Balduvian Bears
[ERROR]   PlayTopCardFromLibraryTest.test_MindsDesire_SplitRightPlay:190->CardTestPlayerAPIImpl.execute:283 Can't find ability to activate command: Cast Revenge$targetPlayer=PlayerB
[ERROR]   StolenGoodsTest.castDualFaceCard:38->CardTestPlayerAPIImpl.execute:283 Can't find ability to activate command: Cast Kolvori, God of Kinship
[ERROR]   ReleaseToTheWindTest.test_Exile_ModalDoubleFacedCard:70->CardTestPlayerAPIImpl.execute:283 Must have playable ability, but not found: Cast Akoum Warrior
[ERROR]   ReleaseToTheWindTest.test_Exile_PermanentCard:34->CardTestPlayerAPIImpl.execute:283 Must have playable ability, but not found: Cast Grizzly Bears
[INFO] 
[ERROR] Tests run: 4146, Failures: 14, Errors: 0, Skipped: 80
```

File: Mage/src/main/java/mage/util/CardUtil.java [hunk 1]
```diff
@@ -1186,8 +1187,24 @@ public static void takeControlUnderPlayerEnd(Game game, Player controller, Playe
...
-    public static void makeCardPlayable(Game game, Ability source, Card card, Duration duration, boolean anyColor) {
-        makeCardPlayable(game, source, card, duration, anyColor, null, null);
+    /**
+     * Groups together the most usual ways a card's payment is adjusted
+     * by card effects that allow play or cast.
+     */
+    public enum SimpleCastManaAdjustment {
+        AS_THOUGH_ANY_MANA_TYPE,
+        AS_THOUGH_ANY_MANA_COLOR,
+        WITHOUT_PAYING_MANA_COST,
```
Comment (xenohedron) [inline Mage/src/main/java/mage/util/CardUtil.java:1197]: I think it would be better to have a `NONE` option in the enum rather than using `null` as an option.

(Also maybe leave out the word "simple" to shorten it.)

**Comment by xenohedron** [comment]
(Not really related to this PR, but goodness `CardUtil` has all sorts of various and sundry things stashed in it.)

**Comment by correl** [comment]
Re: [Nahiri, Forged in Fury](https://scryfall.com/card/mat/36/nahiri-forged-in-fury), while implementing it I noticed that the casting cost manipulation occurs before the choice to cast ever occurs, which I imagine makes the MDFCs difficult to deal with. Also, per the reading of the card, it seemed to me that Forged in Fury's multiple "you may play/cast" clauses overlap in a way that would allow the player to choose whether or not to pay the casting cost for equipment. I could see no way to implement that either, leaving no way to choose to cast equipment like [Sigil of Distinction](https://scryfall.com/card/ala/219/sigil-of-distinction) with its `X` cost paid.

If there's any way I can help or any functionality I overlooked while implementing it, let me know, it's been bugging me that I didn't do more with it.

**Comment by xenohedron** [comment]
> Re: [Nahiri, Forged in Fury](https://scryfall.com/card/mat/36/nahiri-forged-in-fury), while implementing it I noticed that the casting cost manipulation occurs before the choice to cast ever occurs, which I imagine makes the MDFCs difficult to deal with. Also, per the reading of the card, it seemed to me that Forged in Fury's multiple "you may play/cast" clauses overlap in a way that would allow the player to choose whether or not to pay the casting cost for equipment. I could see no way to implement that either, leaving no way to choose to cast equipment like [Sigil of Distinction](https://scryfall.com/card/ala/219/sigil-of-distinction) with its `X` cost paid.
> 
> If there's any way I can help or any functionality I overlooked while implementing it, let me know, it's been bugging me that I didn't do more with it.

I thought your implementation was very clean considering the current constraints of the engine. Currently there's no way to choose which "as though" effect to use (see e.g. #9268, which uncovered another deeper issue, currently stalled). There are also quite a few things broken with double faced cards; this is an interesting issue here with the God/Equipment MDFCs and I really have no idea how much restructuring might be needed to solve it. I think Nahiri ought to create two independent effects to do it properly.

Hmmm... maybe try to model the second effect off Omniscience?

**Comment by Susucre** [comment]
Responding in batch:

>Failing tests for reference.

Definitively expected, as I did not make it yet to handle the WITHOUT_PAYING_MANA_COST case properly in `makePlayableOrCastable`.

> I marked the PR as draft 

Thanks! I did not know of that part of PR settings.

> goodness CardUtil has all sorts of various and sundry things stashed in it.

I did not question `makeCardPlayable` being in CardUtil and went all in moving the `exileAndMakePlayable` there too. Now that it is getting deeper in those effects, I do not mind moving the different functions in this PR in a dedicated file next to it. As a general note, not sure we want to split CardUtil as a general code aesthetic improvment? (Obviously out of scope there.)

> NONE option in the enum rather than using null as an option. Also maybe leave out the word "simple" to shorten it.

Alright, makes sense. 

As for @correy on Nahiri, do not stress too much over that, this is more of "damn mdfc are complicated", and I was looping over all the similar effect, this one stand out as being very slightly bugged in a corner case scenario. I'll end up making unit tests for that eventually, and either fix it in the PR, or make a followup issue. Maybe two calls to  `makeCardPlayableOrCastable` with conditions to affect equipment spells and non equipment spells respectivelly?
I had mdfc in mind as one of the bugged scenario that motivated this rework was relative to double faced cards that are a nonland on the front, and a land in the back. You may cast the front but not the back, and you may play either of them.

**Comment by Susucre** [comment]
Thinking more about it, there is a sub category of cards that apply their asThough conditionally when the spell is cast, and as of now we are doing it wrongly testing the card before applying the asthough effect. Mdfc are so much fun. 

For now I did found 4 of them:
https://scryfall.com/search?q=%21%22Arvinox%2C+the+Mind+Flail%22or%21%22Narset%2C+Enlightened+Master%22or%21%22Chandra%2C+Hope%27s+Beacon%22or%21%22Nivix%2C+Aerie+of+the+Firemind%22&unique=cards&as=grid&order=name

Nahiri could be handled like them (once them work properly) with two `makeCardCastable`, one applying to equipment spells, and the other to non-equipment spells (Or should you be able to cast an equipment for its cost as an option? Then indeed we'd need #9268 to allow for both at the same time)

File: Mage.Sets/src/mage/cards/b/BrainstealerDragon.java [hunk 2]
```diff
@@ -99,7 +99,7 @@ public boolean apply(Game game, Ability source) {
         for (Card card : cards.getCards(game)) {
             CardUtil.makeCardPlayableOrCastable(
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/b/BrainstealerDragon.java:100]: It's a good practice to make two methods for different rules/abilities -- one for "you may play cards" and another for "you may cast spells". With inner usage of shared method like `makeCardPlayableOrCastable`.

**Comment by xenohedron** [comment]
By the way, Oracle text got updated for a bunch of these effects, something to verify. Examples:

[[Vizier of the Menagerie]]
[[Thief of Sanity]]

**Comment by Susucre** [comment]
I'll continue work on this one in the near future.
Two new cards from cmm:
[[Capricious Sliver]]
[[Chandra, Legacy of Fire]]

**Comment by Susucre** [comment]
LTC ones:
[[Lidless Gaze]]
[[Lobelia, Defender of Bag End]]  -- Maybe?

**Comment by xenohedron** [comment]
I'm cleaning up some cards in #10926 (all of those involve casting an instant/sorcery spell from graveyard), in case that impacts this

**Comment by Susucre** [comment]
I'll take a look this evening at your commit. (and start take note of the WOE cards in scope) 

My current plant for the current PR is to  wait for full woe+wot, then give it a good cleanup and rework the effects not yet handled by the CardUtils methods. Probably will accept appliers to add the non standard AsThought modifiers like 'pay with life'. Oh and add unit testing all around. 

It was a big task, and at some point I'd realised I wanted more card implementation experience before refactoring the very complex one.

**Comment by Susucre** [comment]
This has become quite stale...
I still think the rework is worthy, but due to the amount of various templating, I think I'll attempt to redo it in smaller chunks in the future.
